### PR TITLE
[corlib] Pass null-terminated string for logging

### PR DESCRIPTION
**Pull request by luhenry**

### CODE REVIEW

**Comment by lewurm** [comment]
needs backport to `master` and `2018-02`

File: mcs/class/corlib/System.IO/LogcatTextWriter.cs [hunk 1]
```diff
@@ -48,22 +48,52 @@ public override void WriteLine ()
...
+
+		void Log (string message)
+		{
+			const int buffer_size = 512;
+
 			unsafe {
-				fixed (byte *b_appname = appname)
-				fixed (byte *b_message = Encoding.UTF8.GetBytes(o)) {
-					Log (b_appname, 1 << 5 /* G_LOG_LEVEL_MESSAGE */, b_message);
+				if (message.Length < buffer_size) {
```
Comment (jonpryor) [inline mcs/class/corlib/System.IO/LogcatTextWriter.cs]: This is wrong. You don't want `message.Length < buffer_size`, as `message` may contain e.g. emoji, which will take more than one byte to convert.

Instead, you want `if (Encoding.UTF8.GetByteCount (message) < buffer_size)`.

File: mcs/class/corlib/System.IO/LogcatTextWriter.cs [hunk 2]
```diff
@@ -48,22 +48,52 @@ public override void WriteLine ()
...
+				}
+
+				fixed (byte *b_message = Encoding.UTF8.GetBytes(message + '\0')) {
+					Log (b_message);
 				}
 			}
-			stdout.WriteLine (o);
 		}
 
-		public static bool IsRunningOnAndroid ()
```
Comment (jonpryor) [inline mcs/class/corlib/System.IO/LogcatTextWriter.cs]: In terms of the diff itself, it's odd that you moved `IsRunningOnAndroid()` to be *after* all the `Log()` overloads. The patch would be smaller/have less noise if the `Log(byte*)` overload were *after* the "existing" `IsRunningOnAndroid()`.

File: mcs/class/corlib/System.IO/LogcatTextWriter.cs [hunk 3]
```diff
@@ -48,13 +48,43 @@ public override void WriteLine ()
 			var o = line.ToString ();
 			line.Clear ();
 
+			Log (o);
+			stdout.WriteLine (o);
+		}
+
+		void Log (string message)
```
Comment (marek-safar) [inline mcs/class/corlib/System.IO/LogcatTextWriter.cs]: Should be static

File: mcs/class/corlib/System.IO/LogcatTextWriter.cs [hunk 4]
```diff
@@ -48,13 +48,43 @@ public override void WriteLine ()
...
+				}
+
+				fixed (byte *b_message = Encoding.UTF8.GetBytes(message + '\0')) {
+					Log (b_message);
 				}
 			}
-			stdout.WriteLine (o);
+		}
+
+		unsafe void Log (byte* b_message)
```
Comment (marek-safar) [inline mcs/class/corlib/System.IO/LogcatTextWriter.cs]: Should be static

Comment (luhenry) [reply mcs/class/corlib/System.IO/LogcatTextWriter.cs]: It accesses `appname` which is non-static.

File: mcs/class/corlib/System.IO/LogcatTextWriter.cs (hunk 3 above)
Comment (luhenry) [reply mcs/class/corlib/System.IO/LogcatTextWriter.cs]: It calls `Log(byte*)` which can't be static.

**Comment by luhenry** [comment]
@monojenkins backport to 2018-02

**Comment by luhenry** [comment]
@monojenkins backport to master

**Pull request merged**